File: python/crud.py
import asyncio
import uuid

from golem_base_sdk import GolemBaseClient, GolemBaseCreate, Annotation, GolemBaseUpdate, GolemBaseDelete, GenericBytes

GOLEM_DB_RPC = "https://kaolin.holesky.golemdb.io/rpc"
GOLEM_DB_WSS = "wss://kaolin.holesky.golemdb.io/rpc/ws"
PRIVATE_KEY = "<YOUR_PRIVATE_KEY>"

async def crud_example():
    golem_base_client = await GolemBaseClient.create(
        rpc_url=GOLEM_DB_RPC,
        ws_url=GOLEM_DB_WSS,
        private_key=PRIVATE_KEY,
    )
    print("Golem DB client initialized")

    await golem_base_client.watch_logs(
        label="",
        create_callback=lambda create: print(f"WATCH-> Create: {create}"),
        update_callback=lambda update: print(f"WATCH-> Update: {update}"),
        delete_callback=lambda delete: print(f"WATCH-> Delete: {delete}"),
        extend_callback=lambda extend: print(f"WATCH-> Extend: {extend}"),
    )

    # create a new entity with annotations
    id = str(uuid.uuid4())
    receipt = await golem_base_client.create_entities([GolemBaseCreate(
        b'Test entity',
        600,
        [
            Annotation("testTextAnnotation", "demo"), 
            Annotation("id", id),
        ],
        [
            Annotation("version", 1)
        ]
    )])
    print(f"Receipt: {receipt}")

    # query the entity by annotations
    entities = await golem_base_client.query_entities(f'id = "{id}" && version = 1')
    for entity in entities:
        print(f"Entity: {entity}")
    
    # update the entity
    receipt = await golem_base_client.update_entities([GolemBaseUpdate(
        receipt[0].entity_key,
        b'Updated entity',
        1200,
        [
            Annotation("id", id)
        ],
        [
            Annotation("version", 2)
        ]
    )])
    print(f"Receipt: {receipt}")

    # query updated entity by annotations
    entity_key = None
    entities = await golem_base_client.query_entities(f'id = "{id}" && version = 2')
    for entity in entities:
        print(f"Entity: {entity}")
        # we can also obtain the entity key from the entity object
        entity_key = entity.entity_key

    # remove the entity
    receipt = await golem_base_client.delete_entities([GolemBaseDelete(
        GenericBytes.from_hex_string(entity_key),
    )])
    print(f"Receipt: {receipt}")

    await golem_base_client.disconnect()

asyncio.run(crud_example())

File: ts/crud.ts
import { createClient, AccountData, Tagged, GolemBaseCreate, Annotation, GolemBaseUpdate } from 'golem-base-sdk'
import { randomUUID } from 'crypto'

const key: AccountData = new Tagged(
    "privatekey", 
    Buffer.from('<YOUR_PRIVATE_KEY>', 'hex')
)  // TODO: replace with your private key, this variable will be used in the mutations operations
const chainId = 60138453025  // ID of the chain of our public testnet - Kaolin
const rpcUrl = "https://kaolin.holesky.golemdb.io/rpc"  // RPC URL of the chain
const wsUrl = "wss://kaolin.holesky.golemdb.io/rpc/ws"  // WebSocket URL of the chain

// TextEncoder and TextDecoder are used to encode and decode data from text into bytes and vice versa
const encoder = new TextEncoder()
const decoder = new TextDecoder()

// Create a client to interact with the GolemDB API
const client = await createClient(
    chainId,
    key,
    rpcUrl,
    wsUrl,
);

// Watch for logs on the chain
const unsubscribe = client.watchLogs({
    fromBlock: BigInt(await client.getRawClient().httpClient.getBlockNumber()),
    onCreated: (args) => {
        console.log("WATCH-> Create: ", args)
    },
    onUpdated: (args) => {
        console.log("WATCH-> Update: ", args)
    },
    onExtended: (args) => {
        console.log("WATCH-> Extend: ", args)
    },
    onDeleted: (args) => {
        console.log("WATCH-> Delete: ", args)
    },
    onError: (error) => {
        console.error("WATCH-> Error: ", error)
    },
    pollingInterval: 1000,
    transport: "websocket",
})

// Create a new entity with annotations
const id = randomUUID()
const creates = [
    {
      data: encoder.encode("Test entity"),
      btl: 600,
      stringAnnotations: [new Annotation("testTextAnnotation", "demo"), new Annotation("id", id)],
      numericAnnotations: [new Annotation("version", 1)]
    } as GolemBaseCreate]

const createReceipt = await client.createEntities(creates);
console.log('Receipt', createReceipt)

// Query the entity by annotations
let entities = await client.queryEntities(`id = "${id}" && version = 1`)
let entityKey: `0x${string}` | null = null
for (const entity of entities) {
    console.log('Entity value', decoder.decode(entity.storageValue))
    entityKey = entity.entityKey
}

// Update the entity
const updateReceipt = await client.updateEntities([{
    entityKey: createReceipt[0].entityKey,
    data: encoder.encode("Updated entity"),
    btl: 1200,
    stringAnnotations: [new Annotation("id", id)],
    numericAnnotations: [new Annotation("version", 2)]
} as GolemBaseUpdate])
console.log('Update', updateReceipt)

// Query the entity by annotations
entities = await client.queryEntities(`id = "${id}" && version = 2`)
console.log('Entities', entities)

// Delete the entity
const deleteReceipt = await client.deleteEntities([entityKey as `0x${string}`])
console.log('Delete', deleteReceipt)

unsubscribe()
process.exit(0)